core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java

```diff
@@ -542,14 +595,108 @@ public GroupByIdBlock getResult()
+            if (!block.isNull(blockOffset + i)) {
+                groupIds[groupIdOffset + i] = this.groupIds[hashPositions[i]];
+            }
+            else {
+                groupIds[groupIdOffset + i] = NULL_GROUP_ID_MARKER;
+            }
+        }
+
+        for (int i = 0; i < batchSize; i++) {
+            if (groupIds[groupIdOffset + i] >= 0) {
+                long value = BIGINT.getLong(block, blockOffset + i);
+                long storedValue = values[hashPositions[i]];
+                boolean match = value == storedValue;
+                groupIds[groupIdOffset + i] = (groupIds[groupIdOffset + i] + 1) * (match ? 1 : 0) - 1;
+            }
+        }
+
+        if (nullGroupId < 0) { // Assigning new null group id adds another branch inside a loop so we want to do it only once
+            for (int i = 0; i < batchSize; i++) {
+                if (groupIds[groupIdOffset + i] == NULL_GROUP_ID_MARKER) {
+                    if (nullGroupId < 0) {
+                        // set null group id
+                        nullGroupId = nextGroupId++;
+                    }
+                    groupIds[groupIdOffset + i] = nullGroupId;
+                }
+                else if (groupIds[groupIdOffset + i] == -1) {
+                    groupIds[groupIdOffset + i] = putIfAbsent(BIGINT.getLong(block, blockOffset + i), hashPositions[i]);
+                }
+            }
+        }
+        else {
+            for (int i = 0; i < batchSize; i++) {
+                if (groupIds[groupIdOffset + i] == NULL_GROUP_ID_MARKER) {
+                    groupIds[groupIdOffset + i] = nullGroupId;
+                }
+                else if (groupIds[groupIdOffset + i] == -1) {
+                    groupIds[groupIdOffset + i] = putIfAbsent(BIGINT.getLong(block, blockOffset + i), hashPositions[i]);
+                }
+            }
+        }
+    }
+
+    private void batchedPutIfAbsentNoNull(int batchSize, int blockOffset, Block block, long[] groupIds, int groupIdOffset)
+    {
+        int[] hashPositions = new int[batchSize];
+        for (int i = 0; i < batchSize; i++) {
+            hashPositions[i] = getHashPosition(BIGINT.getLong(block, blockOffset + i), mask);
+        }
+
+        for (int i = 0; i < batchSize; i++) {
+            groupIds[groupIdOffset + i] = this.groupIds[hashPositions[i]];
+        }
+
+        for (int i = 0; i < batchSize; i++) {
+            if (groupIds[groupIdOffset + i] != -1) {
+                long value = BIGINT.getLong(block, blockOffset + i);
+                long storedValue = values[hashPositions[i]];
+                boolean match = value == storedValue;
+                // Same as
+                // if (!match)
+                //   groupIds[groupIdOffset + i] = -1;
+                // but without explicit branches
+                groupIds[groupIdOffset + i] = (groupIds[groupIdOffset + i] + 1) * (match ? 1 : 0) - 1;
+            }
+        }
+
+        for (int i = 0; i < batchSize; i++) {
```

Comment:
potential improvement for this loop (may or may not be beneficial).
we could skip it if no new groups (no -1 in the groupIds). This should be the case for aggregations with a small number of groups. To quickly check if there is at least one -1 we could `bitwise or (|)` all groupIds in one of the loops above and compare the result with -1.

Reply:
Tried that at the very beginning. No difference no matter how many groups there are.